src/protocol.rs

```diff
@@ -121,145 +131,70 @@ impl Participant {
         &mut self,
         message: &Message,
         rng: &mut R,
-    ) -> Result<(Identifier, Output, Vec<Message>)> {
+    ) -> Result<(Identifier, Option<P::Output>, Vec<Message>)> {
         info!("Processing single message.");
 
         if message.to() != self.id {
             return Err(InternalError::WrongMessageRecipient);
         }
-        match message.message_type() {
-            MessageType::Auxinfo(_) => {
+        if message.id() != self.sid {
+            error!(
+                "Message for session {} was routed to the wrong participant (sid: {})!",
+                message.id(),
+                self.sid
+            );
+            return Err(InternalError::ProtocolError);
```

Comment:
nit: this and the above "wrong message recipient" should be a "calling application error" instead.